cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -127,6 +124,69 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 	return etcdClient, nil
 }
 
+// retrieveEtcdEndpoints returns the list of etcd endpoints.
+func retrieveEtcdEndpoints(client clientset.Interface) ([]string, error) {
+	etcdEndpoints, err := retrieveEtcdEndpointsFromPodAnnotationWithRetry(client)
+	if err != nil {
+		return retrieveEtcdEndpointsFromClusterStatus(client)
+	}
+	return etcdEndpoints, nil
+}
+
+// retrieveEtcdEndpointsFromPodAnnotation returns the list of endpoints as reported on etcd's pod annotations
+func retrieveEtcdEndpointsFromPodAnnotationWithRetry(client clientset.Interface) ([]string, error) {
+	etcdEndpoints := []string{}
+	var lastErr error
+	err := wait.ExponentialBackoff(clientsetretry.DefaultRetry, func() (bool, error) {
+		if etcdEndpoints, lastErr = retrieveEtcdEndpointsFromPodAnnotation(client); lastErr != nil {
+			return false, nil
+		}
+		// If the list of etcd endpoints is empty we want to retry: this can happen if joining a secondary
+		// control plane while the primary control plane didn't mirror its static pods yet.
+		return len(etcdEndpoints) > 0, nil
+	})
+	if err != nil {
+		if lastErr != nil {
+			return []string{}, errors.Wrap(lastErr, "could not retrieve the list of etcd endpoints")
+		}
+		return []string{}, errors.Wrap(err, "could not retrieve the list of etcd endpoints")
+	}
+	return etcdEndpoints, nil
+}
+
+func retrieveEtcdEndpointsFromPodAnnotation(client clientset.Interface) ([]string, error) {
```

Comment:
`getRawEtcdEndpointsFromPodAnnotation`for consistency with the API endpoint part